actix/src/handler.rs

```diff
@@ -151,6 +151,93 @@ where
     }
 }
 
+/// Response Type that can borrow actor and it's context in async/await.
+/// `ResponseAsync::atomic` would borrow `&mut Actor` and `&mut Actor::Context`.
+/// It would have exclusive access to them and block the actor until the future resolved.
+///
+/// # Examples:
+/// ```rust, no_run
+/// # use actix::prelude::*;
+/// # use actix::clock::sleep;
+/// # use std::time::Duration;
+/// #
+/// # #[derive(Message)]
+/// # #[rtype(result = "usize")]
+/// # struct Msg;
+///
+/// #[derive(Message)]
+/// # #[rtype(result = "usize")]
+/// # struct Msg2;
+/// #
+/// # struct MyActor(usize);
+/// #
+/// # impl Actor for MyActor {
+/// #    type Context = Context<Self>;
+/// # }
+/// #
+/// impl Handler<Msg> for MyActor {
+///     type Result = AsyncResponse<Self, usize>;
+///
+///     fn handle(&mut self, msg: Msg, ctx: &mut Context<Self>) -> Self::Result {
+///         AsyncResponse::atomic(self, ctx, |act, ctx| async move {
+///             let msg = msg;
+///             let ctx = ctx;
+///             act.0 = 30;
+///             sleep(Duration::from_millis(1)).await;
+///             act.0 -= 1;
+///             act.0
+///         })
+///     }
+/// }
+/// ```
+pub enum AsyncResponse<A, T> {
+    #[doc(hidden)]
+    Concurrent(()),
+    Atomic(ResponseActFuture<A, T>),
+}
+
+impl<A, T> AsyncResponse<A, T>
+where
+    A: Actor,
+    A::Context: AsyncContext<A>,
+    T: 'static,
+{
+    pub fn atomic<'a, F, Fut>(act: &'a mut A, ctx: &'a mut A::Context, fut: F) -> Self
+    where
+        F: FnOnce(&'a mut A, &'a mut A::Context) -> Fut + 'static,
+        Fut: Future<Output = T> + 'a,
+    {
+        let fut: Pin<Box<dyn ActorFuture<A, Output = T>>> =
+            Box::pin(crate::fut::wrap_future::<_, A>(fut(act, ctx)));
+        // SAFETY:
+        //
+        // Borrow &mut Actor and &mut Context<Actor> is safe because the actor would be
+        // blocked on wait future until it's resolved.
+        // This is achieved with Context::wait which give exclusive access to actor and context.
+        let fut = unsafe { core::mem::transmute(fut) };
+        AsyncResponse::Atomic(fut)
+    }
+}
+
+impl<A, M, T: 'static> MessageResponse<A, M> for AsyncResponse<A, T>
+where
+    A: Actor,
+    A::Context: AsyncContext<A>,
+    M: Message<Result = T>,
+    T: 'static,
+{
+    fn handle(self, ctx: &mut A::Context, tx: Option<OneshotSender<M::Result>>) {
+        match self {
+            AsyncResponse::Atomic(fut) => {
+                ctx.wait(fut.map(move |res, _, _| tx.send(res)));
```

Comment:
Sorry for reviving this....

Isn't this unsound already ? You have two `&mut A::Context` alive at the same time, one given to the method and the other inside the future.